### PR TITLE
chore(doc): Add comments and derive attributes

### LINKED ISSUE #42: Enhance documentation, add doctest

### LINKED ISSUE #62: Use a faster `HashMap` crate
I don't think hash collision is a concern for this crate. Consider using a faster `HashMap` impl than `std::collections::HashMap`.

Reference:
https://nnethercote.github.io/perf-book/hashing.html
https://nnethercote.github.io/perf-book/linting.html#disallowing-types

### PR DESCRIPTION
**Pull request by malik672**
this is to solve the documentation issue and close issue #50 and

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,11 +2,38 @@
...
 pub mod entities;
+/// Contains error types  for the Uniswap SDK Core.
+///
+/// This module defines custom error types that are used throughout the SDK to
+/// handle various error conditions.
 pub mod error;
+/// Contains commonly used items from the Uniswap SDK Core.
+///
+/// This module re-exports items that are commonly used together,
+/// making it easier to import them in other parts of your application.
```
Comment (shuhuiluo) [inline src/lib.rs:33]: These comments should be placed on the top of each module.

File: src/entities/mod.rs [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
+/// supported by the Uniswap SDK Core.
 pub mod currency;
+/// The `ether` module provides utilities and functionalities specific to Ether, the native
+/// cryptocurrency of the Ethereum blockchain.
 pub mod ether;
+/// The `fractions` module contains functionality related to fractional values, such as currency
+/// amounts, percentages, and prices.
 pub mod fractions;
+/// The `token` module provides definitions and utilities for handling tokens on the Uniswap
+/// platform, including their creation, management, and interactions.
```
Comment (shuhuiluo) [inline src/entities/mod.rs]: The AI generated description is not accurate.

File: src/entities/fractions/mod.rs [hunk 3]
```diff
@@ -1,4 +1,11 @@
+/// This module contains functionality related to currency amounts in the context of fractions.
 pub mod currency_amount;
+/// This module contains functionality related to fractions, which are used for various calculations
+/// in the SDK.
 pub mod fraction;
+/// This module contains functionality related to percentages, which are used for expressing
+/// fractions in a more human-readable format.
 pub mod percent;
+/// This module contains functionality related to prices, which are used for calculating exchange
+/// rates and other financial metrics.
```
Comment (shuhuiluo) [inline src/entities/fractions/mod.rs]: These comments are so AI that they are too general and don't provide much insight.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
Do not base the doc changes on the `big-decimal` changes as it changes the behaviour of this SDK. The amount of comments is excessive especially in the `addresses` mod since most of the variables are self-explanatory.

**Comment by malik672** [comment]
> Do not base the doc changes on the `big-decimal` changes as it changes the behaviour of this SDK. The amount of comments is excessive especially in the `addresses` mod since most of the variables are self-explanatory.

tbh I didn't do it myself, used phind to generate most of the comments

File: src/entities/fractions/mod.rs (hunk 3 above)
Comment (malik672) [reply src/entities/fractions/mod.rs]: true, will still have to change them just wanted something to fill the comments

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
I don't mean to waste your time. But could you start a new PR and only add comments where the code is not self-explanatory?

**Comment by malik672** [comment]
> I don't mean to waste your time. But could you start a new PR and only add comments where the code is not self-explanatory

sure, point me to the parts

**Comment by shuhuiluo** [comment]
> > I don't mean to waste your time. But could you start a new PR and only add comments where the code is not self-explanatory
> 
> sure, point me to the parts

Use your judgement. Break down the changes.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
As I said, start a new PR and only add comments where the code isn't self-explanatory. Do not fill the code with useless AI comments to reduce readability. Move the hash map change to a different PR.

**Comment by malik672** [comment]
> As I said, start a new PR and only add comments where the code isn't self-explanatory. Do not fill the code with useless AI comments to reduce readability. Move the hash map change to a different PR.

yeah true, but there would have to be documentation though even for the self explanatory ones(from alloy) and can it not just be possible to just do both in one pr and maybe just change the title

**Comment by shuhuiluo** [comment]
> > As I said, start a new PR and only add comments where the code isn't self-explanatory. Do not fill the code with useless AI comments to reduce readability. Move the hash map change to a different PR.
> 
> yeah true, but there would have to be documentation though even for the self explanatory ones(from alloy) and can it not just be possible to just do both in one pr and maybe just change the title

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/best-practices-for-pull-requests

File: src/addresses.rs [hunk 4]
```diff
@@ -163,6 +176,7 @@ lazy_static! {
 }
 
 lazy_static! {
+    /// The `MAINNET_ADDRESSES` struct holds the Uniswap contract addresses for the Goerli Testnet.
```
Comment (shuhuiluo) [inline src/addresses.rs]: Typo.

File: src/addresses.rs [hunk 5]
```diff
@@ -4,7 +4,9 @@ type AddressMap = HashMap<u64, Address>;
 type ChainMap = HashMap<u64, ChainAddresses>;
 type ChainAddress = HashMap<u64, Address>;
 
-#[derive(Clone, Copy)]
+#[derive(Clone, Copy, Debug)]
+
+/// Represents the addresses of various core contracts of uniswap on a network.
```
Comment (shuhuiluo) [inline src/addresses.rs]:
```suggestion
/// Represents the addresses of various core contracts of uniswap on a network.
#[derive(Clone, Copy, Debug)]
```
Put derive attributes next to the struct.

File: src/chains.rs [hunk 6]
```diff
@@ -1,29 +1,58 @@
 #[derive(Debug, Clone, Copy)]
+/// Represents the unique identifier for different blockchain networks supported by the Uniswap SDK.
+///
+/// Each variant corresponds to a specific blockchain network, identified by its unique chain ID.
```
Comment (shuhuiluo) [inline src/chains.rs]:
```suggestion
/// Represents the unique identifier for different blockchain networks supported by the Uniswap SDK.
///
/// Each variant corresponds to a specific blockchain network, identified by its unique chain ID.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

File: src/chains.rs [hunk 7]
```diff
@@ -47,13 +76,23 @@ pub const SUPPORTED_CHAINS: [ChainId; 20] = [
     ChainId::ROOTSTOCK,
 ];
 
+#[derive(Debug, Clone, Copy)]
+/// Represents the names of native currencies supported by the Uniswap SDK.
```
Comment (shuhuiluo) [inline src/chains.rs]:
```suggestion
/// Represents the names of native currencies supported by the Uniswap SDK.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

File: src/constants.rs [hunk 8]
```diff
@@ -3,19 +3,31 @@ use alloy_primitives::U256;
 use num_bigint::Sign;
 
 #[derive(Clone, Copy, Debug, PartialEq)]
+/// Represents the various type of trades.
```
Comment (shuhuiluo) [inline src/constants.rs]:
```suggestion
/// Represents the various type of trades.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

File: src/constants.rs [hunk 9]
```diff
@@ -3,19 +3,31 @@ use alloy_primitives::U256;
...
     ExactInput,
+
+    /// Indicates that the trade is based on an exact output amount.
     ExactOutput,
 }
 
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq, Copy)]
+
+/// Represents three various way to rounds
```
Comment (shuhuiluo) [inline src/constants.rs]:
```suggestion
/// Represents three various way to rounds
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```

File: src/entities/base_currency.rs [hunk 10]
```diff
@@ -3,11 +3,23 @@ use std::ops::Deref;
 use alloy_primitives::ChainId;
 
 #[derive(Clone, PartialEq, Debug)]
+
+/// `CurrencyLike` is a generic struct representing a currency with a specific chain ID,
+/// decimals, symbol, name, and additional metadata.
```
Comment (shuhuiluo) [inline src/entities/base_currency.rs]:
```suggestion
/// `CurrencyLike` is a generic struct representing a currency with a specific chain ID,
/// decimals, symbol, name, and additional metadata.
#[derive(Clone, PartialEq, Debug)]
```

File: src/entities/currency.rs [hunk 11]
```diff
@@ -1,39 +1,38 @@
 use crate::prelude::*;
 
 #[derive(Clone, PartialEq, Debug)]
+
+/// This enum represnets the two type of currencies i.e native and Token
```
Comment (shuhuiluo) [inline src/entities/currency.rs]:
```suggestion
/// This enum represnets the two type of currencies i.e native and Token
#[derive(Clone, PartialEq, Debug)]
```

File: src/entities/fractions/currency_amount.rs [hunk 12]
```diff
@@ -118,6 +120,11 @@ impl<T: CurrencyTrait> CurrencyAmount<T> {
             return Err(Error::NotEqual());
         }
 
+        if decimal_places == 0 {
+            // Directly convert the numerator to a string for zero decimal places
+            return Ok(self.numerator().to_string());
+        }
+
```
Comment (shuhuiluo) [inline src/entities/fractions/currency_amount.rs:127]:
```suggestion
```
That's wrong.

File: src/entities/fractions/fraction.rs [hunk 13]
```diff
@@ -2,11 +2,12 @@
 use crate::prelude::*;
 use std::ops::{Add, Deref, Mul, Sub};
 
-/// Struct representing a fraction with metadata
+/// struct to represent a fraction
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs]:
```suggestion
/// Struct representing a fraction with metadata
```

File: src/entities/fractions/fraction.rs [hunk 14]
```diff
@@ -68,7 +82,7 @@ pub trait FractionBase<M>: Sized {
     /// Accessor method for retrieving metadata
     fn meta(&self) -> M;
 
-    /// Accessor method for retrieving the numerator
+    /// Accessor method for retrieving metadata
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs]:
```suggestion
    /// Accessor method for retrieving the numerator
```

File: src/entities/fractions/fraction.rs [hunk 15]
```diff
@@ -88,7 +102,7 @@ pub trait FractionBase<M>: Sized {
         )
     }
 
-    /// Returns the inverted fraction
+    /// Returns the inverted [`fraction`]
```
Comment (shuhuiluo) [inline src/entities/fractions/fraction.rs]:
```suggestion
    /// Returns the inverted fraction
```

File: src/entities/token.rs [hunk 16]
```diff
@@ -4,9 +4,14 @@ use crate::prelude::*;
 pub type Token = CurrencyLike<TokenMeta>;
 
 #[derive(Clone, PartialEq, Debug)]
+
+/// Represents the metadata for an ERC20 token, including its address and optional fees.
```
Comment (shuhuiluo) [inline src/entities/token.rs]:
```suggestion
/// Represents the metadata for an ERC20 token, including its address and optional fees.
#[derive(Clone, PartialEq, Debug)]
```

File: src/entities/token.rs [hunk 17]
```diff
@@ -40,6 +45,25 @@ impl CurrencyTrait for Token {
 }
 
 impl Token {
+    /// Creates a new `Token` with the given parameters.
```
Comment (shuhuiluo) [inline src/entities/token.rs]:
```suggestion
    /// Creates a new [`Token`] with the given parameters.
```

File: src/entities/token.rs [hunk 18]
```diff
@@ -40,6 +45,25 @@ impl CurrencyTrait for Token {
...
+    /// * `address` - The address of the token.
+    /// * `decimals` - The number of decimals the token uses.
+    /// * `symbol` - The symbol of the token, if any.
+    /// * `name` - The name of the token, if any.
+    /// * `buy_fee_bps` - The buy fee in basis points (bps), if any.
+    /// * `sell_fee_bps` - The sell fee in basis points (bps), if any.
+    ///
+    /// # Returns
+    ///
+    /// A new `Token` instance.
```
Comment (shuhuiluo) [inline src/entities/token.rs]:
```suggestion
    /// A new [`Token`] instance.
```

File: src/error.rs [hunk 19]
```diff
@@ -1,25 +1,25 @@
...
-#[derive(Debug, Error)]
+#[derive(Debug, Error, Clone, Copy)]
 pub enum Error {
-    /// Indicates a mismatch in chain IDs.
+    /// Triggers when compared Chain Ids do not match
     #[error("Chain IDs do not match: {0} and {1}")]
     ChainIdMismatch(u64, u64),
 
-    /// Indicates that two addresses are equal.
+    /// Triggers when comapred addresses are not the smae
```
Comment (shuhuiluo) [inline src/error.rs]:
```suggestion
    /// Triggers when comapred addresses are the smae
```

File: src/error.rs [hunk 20]
```diff
@@ -42,7 +42,7 @@ mod tests {
         assert_eq!(error.to_string(), "Addresses are equal");
     }
 
-    /// Test that `Error::MaxUint` displays the correct error message.
+    /// Test that `Error::MaxUint` displays the correct error messagit push ge.
```
Comment (shuhuiluo) [inline src/error.rs]: Typo.

File: src/utils/mod.rs [hunk 21]
```diff
@@ -1,5 +1,9 @@
+/// Module that contains various methods to comput price
 pub mod compute_price_impact;
+
+/// Module that contains various methods to sort
 pub mod sorted_insert;
+/// Module that contains various methods to compute square root
```
Comment (shuhuiluo) [inline src/utils/mod.rs]:
```suggestion
pub mod compute_price_impact;
pub mod sorted_insert;
```
Aren't even accurate.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/error.rs [hunk 22]
```diff
@@ -42,7 +42,7 @@ mod tests {
         assert_eq!(error.to_string(), "Addresses are equal");
     }
 
-    /// Test that `Error::MaxUint` displays the correct error messagit push ge.
+    /// Test that `Error::MaxUint` displays the correct error messagit.
```
Comment (shuhuiluo) [inline src/error.rs]:
```suggestion
    /// Test that `Error::MaxUint` displays the correct error message.
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Review by shuhuiluo — APPROVED** [review]

**Pull request merged**